File: 백준/Gold/2636. 치즈/치즈.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.StringTokenizer;

class Point {
    int x, y;

    Point(int x, int y) {
        this.x = x;
        this.y = y;
    }

}

public class Main {

    static int n, m;
    static int[][] g;
    static int cnt, time, res;
    static int[] dx = {1, -1, 0, 0};
    static int[] dy = {0, 0, 1, -1};
    static boolean[][] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        g = new int[n][m];
        visited = new boolean[n][m];
        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                g[i][j] = Integer.parseInt(st.nextToken());
                if (g[i][j] == 1) cnt += 1;  // 치즈 개수 증가시키기
            }
        }

        BFS(0, 0);
        System.out.println(time);
        System.out.println(res);
    }

    public static void BFS(int x, int y) {
        visited[x][y] = true;  // 시작점 방문처리
        Deque<Point> dq = new ArrayDeque<>();
        dq.offer(new Point(x, y));

        int meltCnt = 0;
        time += 1;
        while (!dq.isEmpty()) {
            Point now = dq.poll();
            for (int i = 0; i < 4; i++) {
                int nx = now.x + dx[i];
                int ny = now.y + dy[i];
                if (isInner(nx, ny) && !visited[nx][ny]) {
                    if (g[nx][ny] == 1) {   // 치즈 녹여야함
                        visited[nx][ny] = true;
                        g[nx][ny] = 0;
                        meltCnt += 1;  // 녹인 개수 증가
                    } else if (g[nx][ny] == 0) {   // 치즈 없는 곳이면 치즈를 찾기 위해 반복
                        visited[nx][ny] = true;
                        dq.offer(new Point(nx, ny));
                    }
                }
            }
        }
        cnt -= meltCnt;  // 매 횟수마다 녹인 개수 빼줘
        if (cnt == 0) {  // 다 녹인 경우
            res = meltCnt;   // 다 녹였을 때의 마지막 개수 저장
        } else {
            visited = new boolean[n][m];
            BFS(0, 0);  // 다시 탐색 진행
        }
    }

    public static boolean isInner(int x, int y) {
        if (x >= 0 && x < n && y >= 0 && y < m) {
            return true;
        }
        return false;
    }
}

File: 백준/Gold/2636. 치즈/치즈.py
import sys
from collections import deque

# 공기와 접촉된 칸은 한 시간이 지나면 녹아 없어짐
# 공기의 관점에서 생각의 전환 !!
# 공기 중에서 치즈가 모두 녹아 없어지는 데 걸리는 시간,
# 모두 녹기 한 시간 전에 남아있는 치즈조각이 놓여 있는 칸의 개수


n,m = map(int, input().split())
g = [list(map(int, input().split())) for _ in range(n)]

# 0은 공기, 1은 치즈
air = 0
cheeze = 1
dx = [-1,0,1,0]
dy = [0,1,0,-1]
def isInner(x,y):
    if 0<=x<n and 0<=y<m:
        return True
    return False

def BFS(startX,startY):
    dq = deque()
    dq.append((startX,startY))
    visited = [[False] * m for _ in range(n)]
    visited[startX][startY] = True
    cnt = 0 # 치즈 개수
    melt = deque()
    while dq:
        x,y = dq.popleft()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if not isInner(nx,ny): continue
            if visited[nx][ny]: continue
            if g[nx][ny] == air:
                visited[nx][ny] = True
                dq.append((nx,ny))
            elif g[nx][ny] == cheeze:
                visited[nx][ny] = True
                melt.append((nx,ny))
                cnt += 1

    while melt:
        x,y = melt.popleft()
        g[x][y] = air # 공기화
    return cnt




last_cnt = 0
time = 0
while True:

    cnt = BFS(0,0)
    if cnt == 0: # 끝
         break
    time += 1
    last_cnt = cnt

print(time)
print(last_cnt)